students/zach_thomson/lesson07/html_render.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+"""
+
+
+# This is the framework for the base class
+class Element(object):
+
+    tag = 'html'
+
+    indent = '    '
+
+    def __init__(self, content=None, **kwargs):
+        self.style = dict(kwargs)
+        if content is not None:
+            self.contents = [content]
+        else:
+            self.contents = []
+
+    def append(self, new_content):
+        self.contents.append(new_content)
+
+    def _opentag(self):
+        open_tag = ['<{}'.format(self.tag)]
+        for key, value in self.style.items():
+            open_tag.append(' {}="{}"'.format(key, value))
+        open_tag.append('>')
+        return "".join(open_tag)
+
+    def _closetag(self):
+        return "</{}>".format(self.tag)
+
+    def render(self, out_file, cur_ind=''):
+        out_file.write(cur_ind + self._opentag())
+        out_file.write('\n')
+        for content in self.contents:
+            try:
+                content.render(out_file, cur_ind + self.indent)
+            except AttributeError:
+                out_file.write(cur_ind + self.indent + content)
+                out_file.write("\n")
+        out_file.write(cur_ind + self._closetag())
+        out_file.write('\n')
+
+
+
+class Body(Element):
+    tag = 'body'
+
+
+class Html(Element):
+    tag = 'html'
+
+    def render(self, out_file, cur_ind=''):
+        out_file.write("<!DOCTYPE html>\n")
```

Comment:
answered you in chat that yes this would support indentation as well - but not a big deal.